presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveStorageFormats.java

```diff
@@ -430,6 +434,83 @@ private void runTimestampQueries(String tableName, List<TimestampAndPrecision> d
+
+        onPresto().executeQuery(format(
+                "CREATE TABLE %s (\n"
+                        + "id INTEGER\n"
+                        + ",arr ARRAY(TIMESTAMP)\n"
+                        + ",map MAP(TIMESTAMP, TIMESTAMP)\n"
+                        + ",row ROW(col TIMESTAMP)\n"
+                        + ",nested ARRAY(MAP(TIMESTAMP, ROW(col ARRAY(TIMESTAMP))))\n"
+                        + ") WITH (%s)",
+                tableName,
+                format.getStoragePropertiesAsSql()));
+
+        // Insert in a loop because inserting with UNION ALL sometimes makes values invisible to Presto
+        for (TimestampAndPrecision entry : TIMESTAMPS_FROM_HIVE) {
+            onHive().executeQuery(format(
+                    "INSERT INTO %1$s\n"
+                            // insert with SELECT because hive bans array/map/struct functions in VALUES
+                            + "SELECT\n"
+                            + "%3$s\n"
+                            + ",array(%2$s)\n"
+                            + ",map(%2$s, %2$s)\n"
+                            + ",named_struct('col', %2$s)\n"
+                            + ",array(map(%2$s, named_struct('col', array(%2$s))))\n"
+                            // some hive versions don't allow INSERT from SELECT without FROM
+                            + "FROM dummy\n",
+                    tableName,
+                    format("TIMESTAMP '%s'", entry.getWriteValue()),
+                    entry.getId()));
+        }
+
+        for (HiveTimestampPrecision precision : HiveTimestampPrecision.values()) {
+            setSessionProperty(onPresto().getConnection(), "hive.timestamp_precision", precision.name());
+            assertThat(onPresto()
+                    .executeQuery(format(
+                            "SELECT\n"
+                                    + "id\n"
+                                    + ",arr[1]\n"
+                                    + ",map_entries(map)[1][1]\n" // key
+                                    + ",map_entries(map)[1][2]\n" // value
+                                    + ",row.col\n"
+                                    + ",map_entries(nested[1])[1][1]\n" // key
+                                    + ",map_entries(nested[1])[1][2].col[1]\n" // value
+                                    + "FROM %s\n"
+                                    + "ORDER BY id\n",
+                            tableName)))
+                    .as("timestamp containers on %s", format.getName().toLowerCase(Locale.ENGLISH))
+                    .containsExactly(TIMESTAMPS_FROM_HIVE.stream()
+                            .sorted(comparingInt(TimestampAndPrecision::getId))
+                            .map(e -> new Row(Lists.asList(
+                                    e.getId(),
+                                    nCopies(6, Timestamp.valueOf(e.getReadValue(precision))).toArray())))
```

Comment:
The JDBC representation may or may not play tricks on us.
Would be nice to _also_ (first) validate the `typeof(the_array)` and `CAST(the_array AS varchar)`.

See also #6297. I use this technique in some other upcoming (unrelated) PR as well.

Reply:
> some other upcoming (unrelated) PR as well.

see https://github.com/prestosql/presto/pull/6300, the `Verify bound parameters on the server side` commit

Reply:
The type check works fine, but trying to cast these values to varchar like in `runTimestampQueries` causes a Presto compiler error. I don't think this is the kind of bug I was supposed to catch with that test, but it's what I found.

```
SET SESSION hive.timestamp_precision = 'NANOSECONDS';
CREATE TABLE test (col ARRAY(TIMESTAMP));
SELECT CAST(col AS VARCHAR) FROM test;
```
```
Query 20201214_224143_00038_sipnt failed: Compiler failed
```

This is the root cause on the server:
```
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/prestosql/$gen/CursorProcessor_20201214_34.project_0(ConnectorSession;RecordCursor;BlockBuilder;) @75: invokedynamic
  Reason:
    Type 'java/lang/Object' (current frame, stack[1]) is not assignable to 'io/prestosql/spi/type/LongTimestamp'
```

<details><summary>Here is a more-complete version of the stack trace</summary>

```
io.prestosql.spi.PrestoException: Compiler failed
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1386)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:1265)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitProject(LocalExecutionPlanner.java:744)
	at io.prestosql.sql.planner.plan.ProjectNode.accept(ProjectNode.java:82)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:484)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:406)
	at io.prestosql.execution.SqlTaskExecutionFactory.create(SqlTaskExecutionFactory.java:84)
	at io.prestosql.execution.SqlTask.updateTask(SqlTask.java:418)
	at io.prestosql.execution.SqlTaskManager.updateTask(SqlTaskManager.java:395)
	at io.prestosql.server.TaskResource.createOrUpdateTask(TaskResource.java:132)
	at jdk.internal.reflect.GeneratedMethodAccessor748.invoke(Unknown Source)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
        [...] (lots of Jersey and Jetty internals with a few Airtlift lines)]
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: java.lang.VerifyError: Bad type on operand stack
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
	at io.prestosql.sql.gen.ExpressionCompiler.compileCursorProcessor(ExpressionCompiler.java:83)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1349)
	... 71 more
Caused by: java.lang.RuntimeException: java.lang.VerifyError: Bad type on operand stack
	at io.airlift.bytecode.ClassGenerator.defineClasses(ClassGenerator.java:181)
	at io.airlift.bytecode.ClassGenerator.defineClass(ClassGenerator.java:117)
	at io.prestosql.util.CompilerUtils.defineClass(CompilerUtils.java:63)
	at io.prestosql.util.CompilerUtils.defineClass(CompilerUtils.java:57)
	at io.prestosql.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:167)
	at io.prestosql.sql.gen.ExpressionCompiler.compile(ExpressionCompiler.java:134)
	at io.prestosql.sql.gen.ExpressionCompiler.lambda$new$0(ExpressionCompiler.java:70)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 77 more
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/prestosql/$gen/CursorProcessor_20201214_224144_34.project_0(Lio/prestosql/spi/connector/ConnectorSession;Lio/prestosql/spi/connector/RecordCursor;Lio/prestosql/spi/block/BlockBuilder;)V @75: invokedynamic
  Reason:
    Type 'java/lang/Object' (current frame, stack[1]) is not assignable to 'io/prestosql/spi/type/LongTimestamp'
  Current Frame:
    bci: @75
    flags: { }
    locals: { 'io/prestosql/$gen/CursorProcessor_20201214_224144_34', 'io/prestosql/spi/connector/ConnectorSession', 'io/prestosql/spi/connector/RecordCursor', 'io/prestosql/spi/block/BlockBuilder', integer }
    stack: { 'io/prestosql/spi/block/BlockBuilder', 'java/lang/Object' }
  Bytecode:
    0000000: 0336 042d 2c03 b900 4402 0099 000a 0436
    0000010: 0401 a700 0d2c 03b9 0048 0200 c000 4a15
    0000020: 0499 0008 5701 a700 1b0a 1504 9900 0958
    0000030: 5701 a700 0fba 0057 0000 59c7 0006 0436
    0000040: 0415 0499 0008 5701 a700 08ba 005d 0000
    0000050: 1504 9900 0d57 b900 6301 0057 a700 153a
    0000060: 053a 06ba 0069 0000 1906 1905 b900 6f03
    0000070: 00b1                                   
  Stackmap Table:
    full_frame(@21,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95]})
    full_frame(@31,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#74]})
    full_frame(@41,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#74]})
    full_frame(@53,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#74],Long})
    full_frame(@65,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#4]})
    full_frame(@75,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#4]})
    full_frame(@80,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#118]})
    full_frame(@95,{Object[#2],Object[#64],Object[#23],Object[#95],Integer},{Object[#95],Object[#118]})
    same_frame(@113)

	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at com.google.common.reflect.Reflection.initialize(Reflection.java:65)
	at io.airlift.bytecode.ClassGenerator.defineClasses(ClassGenerator.java:177)
	... 88 more
```
</details>

Reply:
See https://github.com/prestosql/presto/issues/4824 for previous cause.

Reply:
Opened as issue #6350.